docgen/props/run.ts

```diff
@@ -1,50 +1,80 @@
-import fs from "fs";
+import { readdir, readFile, writeFile } from "fs/promises";
 import path from "path";
 import { fileURLToPath } from "url";
-import parseInterface from "./parse-interface.js";
-import prettier from "prettier";
-import parseType from "../types/parseTypes.js";
-
-async function formatCode(code: string) {
-  const options = await prettier.resolveConfig(path.join(process.cwd(), ".prettierrc"));
-  return prettier.format(code, { ...options, parser: "typescript" });
-}
+import getInfo from "./getInfo.js";
+import WorkerManager from "./WorkerManager.js";
+import type { WorkerTask } from "./WorkerManager.js";
 
 const dirname = path.dirname(fileURLToPath(import.meta.url));
 const rootDir = path.resolve(dirname, "../../src/lib/components");
-const docTypes = path.resolve(dirname, "../../src/lib/utils/types.json");
+const docTypesPath = path.resolve(dirname, "../../src/lib/utils/types.json");
+let didUpdateAllFiles = true;
 
-async function run() {
-  const componentDirs = fs
-    .readdirSync(rootDir, { withFileTypes: true })
-    .filter((dirent) => dirent.isDirectory())
-    .map((dirent) => dirent.name);
+async function getComponentDirs(rootDir: string) {
+  const dirents = await readdir(rootDir, { withFileTypes: true });
+  return dirents.filter((dirent) => dirent.isDirectory()).map((dirent) => dirent.name);
+}
 
-  for (const dir of componentDirs) {
-    const propsFilePath = path.resolve(rootDir, dir, "props.ts");
+async function updateDocTypesFile(newTypes: Record<string, string>) {
+  let oldTypes: Record<string, string> = {};
 
-    if (fs.existsSync(propsFilePath)) {
-      console.log("processing", propsFilePath);
-      const parsedInterface = parseInterface(propsFilePath);
-      parseType(propsFilePath, docTypes);
+  // if only a few files were updated, we need to build on the existing types data
+  if (!didUpdateAllFiles) {
+    const oldTypesStr = await readFile(docTypesPath, "utf8");
+    oldTypes = JSON.parse(oldTypesStr);
+  }
 
-      let contents = "";
+  const allTypesRaw = Object.assign(oldTypes, newTypes);
 
-      Object.keys(parsedInterface).forEach((varName) => {
-        const interfaceResults = parsedInterface[varName];
+  // sort alphabetically
+  const allTypesSorted = Object.keys(allTypesRaw)
+    .sort()
+    .reduce(
+      (carry, key) => {
+        carry[key] = allTypesRaw[key];
+        return carry;
+      },
+      {} as Record<string, string>
+    );
```

Comment:
Please don't use weird `reduce` functions for that :
```js
const allTypesSorted = Object.fromEntries(
  Object.entries(allTypesRaw).sort((a, b) => a[0].localeCompare(b[0]))
)
```

If you want to ignore accents and case, use `a.localeCompare(b, "en", { sensitivity: "base" })`

Reply:
Thanks, will consider.